### DISCUSSION TITLE
Third-party plugins support

### DISCUSSION THREAD

**Discussion by Sama34** [RevertIT/mybb-rt_discord_webhooks#4] [Ideas]
I did check the source code and I could add support for this plugin by using core hooks, but I think it would be a nice addition to add support for third-party webhook types.

For instance, I would like to add support to my Awards plugin for sending webhooks whenever a new award is granted. It would be nice if there were some hooks to do this easily, specially in the Admin side as the granting logic would be on charge of the third-party plugin.

It is possible I miss some or that the placement isn't necessarily accurate but the following is a list of places where I believe would be necessary to meet what I suggest :

- https://github.com/RevertIT/mybb-rt_discord_webhooks/blob/e06f251a9af5dff28f9d0bdee0bdb1bd5944922f/inc/plugins/rt/src/DiscordWebhooks/Discord/AdminWebhooksAdd.php#L130
- https://github.com/RevertIT/mybb-rt_discord_webhooks/blob/e06f251a9af5dff28f9d0bdee0bdb1bd5944922f/inc/plugins/rt/src/DiscordWebhooks/Discord/AdminWebhooksAdd.php#L159
- https://github.com/RevertIT/mybb-rt_discord_webhooks/blob/e06f251a9af5dff28f9d0bdee0bdb1bd5944922f/inc/plugins/rt/src/DiscordWebhooks/Discord/AdminWebhooksEdit.php#L127
- https://github.com/RevertIT/mybb-rt_discord_webhooks/blob/e06f251a9af5dff28f9d0bdee0bdb1bd5944922f/inc/plugins/rt/src/DiscordWebhooks/Discord/AdminWebhooksEdit.php#L170
- https://github.com/RevertIT/mybb-rt_discord_webhooks/blob/e06f251a9af5dff28f9d0bdee0bdb1bd5944922f/inc/plugins/rt/src/DiscordWebhooks/Discord/AdminWebhooksView.php#L136-L137

**Comment by RevertIT** [comment]
Test out the newest commits as there is no reason to hook onto the backend as it would only complicate things for plugin developers.

You should be able to hook into the `start/end` position of all front-end hooks.

**Reply by Sama34** [reply]
The hooks look nice, but they won't help developers to add new webhook types, which was the aim of my request.

If I wanted to add new webhook types I would still need to hook at `admin_formcontainer_output_row` for adding my plugin's row and some dirty workaround for the insert/update processes.

Please note that my suggestion is aimed at the specific case of pushing webhooks when new awards are granted, which has no relation to the current webhooks offered by the core plugin.

**Reply by RevertIT** [reply]
Noted. See the post below.

**Comment by RevertIT** [comment]
Added a new PR that will let developers hooks anywhere with their plugin.

Hooks are: `rt_discord_webhooks_thirdparty_start` and `rt_discord_webhooks_thirdparty_end` 

Commit: https://github.com/RevertIT/mybb-rt_discord_webhooks/commit/44523de53f90240976339220b1608ed5cccabde9

This hooks on global_start which means now developers can call it anywhere they see it fit.

For example, adding a new hook will be like:

```php
$plugins->add_hook('rt_discord_webhooks_thirdparty_start', 'rt_disc_integration');

function rt_disc_integration()
{
    global $hook_data;

    $hook_data = [
        'webhook_url' => '', // Webhook url from the Discord channel (Set into your plugin a text field where user will be able to set Webhook URL for your plugin)
        'webhook_method' => 'POST', // Webhook method when sending request to API
        'bot' => [
            'username' => '', // Bot username
            'avatar_url' => '', // Bot image url
        ],
        'author' => [
            'name' => '', // Embeds author name
            'url' => '', // Embeds author profile url
            'icon_url' => '' // Embeds author avatar url
        ],
        'title' => '', // Embeds title of the content
        'url' => '', // Embeds url for the title
        'description' => '', // Embeds description
        'char_limit' => 2000, // Embeds max character to be parsed
        'embeds_enabled' => true, // Enable / Disable embeds
        'allow_mentions' => true, // Allow mentions in embeds (@here, @all)
        'color' => '', // HEX embeds color border (eg. #ff0000)
        'timestamp' => (new DateTimeImmutable('@' . TIME_NOW))->format('Y-m-d\TH:i:s\Z'), // Timestamp of the post
        'thumbnail' => [
            'url' => '', // Embeds thumbnail image url
        ],
        'footer' => [
            'text' => '', // Embeds footer text
            'icon_url' => '' // Embeds small footer image url
        ],
        'image' => [
            'url' => '', // Embeds image url
        ]
    ];
}
```
